### PR TITLE
osmomath: QuoTruncate with no reallocation

### PR DESCRIPTION
**Pull request by hieuvubk**
Closes: #6370 

## What is the purpose of the change

QuoTruncate with no reallocation

## Benchmark
```
OLD:
BenchmarkQuoTruncate-16          1781227               663.5 ns/op           448 B/op          6 allocs/op

NEW:
BenchmarkQuoTruncate-16         44712544                24.01 ns/op            0 B/op          0 allocs/op
```

## Testing and Verifying

*(Please pick one of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added unit test that validates ...*
  - *Added integration tests for end-to-end deployment with ...*
  - *Extended integration test for ...*
  - *Manually verified the change by ...*

## Documentation and Release Note

  - [ ] Does this pull request introduce a new feature or user-facing behavior changes?
  - [ ] Changelog entry added to `Unreleased` section of `CHANGELOG.md`?

Where is the change documented? 
  - [ ] Specification (`x/{module}/README.md`)
  - [ ] Osmosis documentation site
  - [ ] Code comments?
  - [ ] N/A

### CODE REVIEW

File: osmomath/decimal.go [hunk 1]
```diff
@@ -386,6 +387,19 @@ func (d BigDec) QuoTruncate(d2 BigDec) BigDec {
 	return BigDec{chopped}
 }
 
+// quotient truncate (mutative)
+func (d BigDec) QuoTruncateMut(d2 BigDec) BigDec {
```
Comment (p0mvn) [inline osmomath/decimal.go:391]: Please add tests for this

**Review by p0mvn — CHANGES_REQUESTED** [review]

File: osmomath/decimal.go (hunk 1 above)
Comment (hieuvubk) [reply osmomath/decimal.go:391]: Added thanks

File: osmomath/decimal_test.go [hunk 2]
```diff
@@ -1560,3 +1560,62 @@ func (s *decimalTestSuite) TestQuoRoundUp_MutativeAndNonMutative() {
...
+			nonMutResult := copy.QuoTruncate(tc.d2)
+
+			// Return is as expected
+			s.Require().Equal(tc.expQuoTruncateMut, nonMutResult, "exp %v, res %v, tc %d", tc.expQuoTruncateMut.String(), tc.d1.String(), tcIndex)
+
+			// Receiver is not mutated
+			s.Require().Equal(tc.d1, copy, "exp %v, res %v, tc %d", tc.expQuoTruncateMut.String(), tc.d1.String(), tcIndex)
+
+			// Receiver is mutated.
+			tc.d1.QuoTruncateMut(tc.d2)
```
Comment (p0mvn) [inline osmomath/decimal_test.go:1614]: Nice, thanks for adding these!

File: x/concentrated-liquidity/math/math.go [hunk 3]
```diff
@@ -88,8 +88,7 @@ func CalcAmount0Delta(liq, sqrtPriceA, sqrtPriceB osmomath.BigDec, roundUp bool)
 	// Each intermediary step is truncated at precision end to get a smaller final amount.
 	// Note that the order of divisions is important here. First, we divide by a larger number (sqrtPriceB) and then by a smaller number (sqrtPriceA).
 	// This leads to a smaller error amplification.
-	// TODO (perf): QuoTruncate with no reallocation.
-	return liq.MulTruncate(diff).QuoTruncate(sqrtPriceB).QuoTruncate(sqrtPriceA)
+	return liq.MulTruncate(diff).QuoTruncateMut(sqrtPriceB).QuoTruncateMut(sqrtPriceA)
```
Comment (p0mvn) [inline x/concentrated-liquidity/math/math.go:91]: Are there any other examples where we can apply these?

**Review by p0mvn — APPROVED** [review]

File: x/concentrated-liquidity/math/math.go [hunk 4]
```diff
@@ -166,7 +166,7 @@ func GetNextSqrtPriceFromAmount0OutRoundingUp(sqrtPriceCurrent, liquidity, amoun
 // avoid overpaying out of the pool. Therefore, we round down.
 // sqrt_next = sqrt_cur + token_in / liq
 func GetNextSqrtPriceFromAmount1InRoundingDown(sqrtPriceCurrent, liquidity, amountOneRemainingIn osmomath.BigDec) (sqrtPriceNext osmomath.BigDec) {
-	return sqrtPriceCurrent.Add(amountOneRemainingIn.QuoTruncate(liquidity))
+	return sqrtPriceCurrent.Add(amountOneRemainingIn.QuoTruncateMut(liquidity))
```
Comment (p0mvn) [inline x/concentrated-liquidity/math/math.go:169]: I don't think we can do this one. Can't mutate `amountOneRemainingIn`

Comment (hieuvubk) [reply x/concentrated-liquidity/math/math.go:169]: Yeah. u right, revert now

**Comment by p0mvn** [comment]
@hieuvubk please resolve one remaining conflict and we can merge this. Sorry for the break again - I'm merging a lot today

I reviewed calls to `QuoTruncate` and agree that we cannot optimize anything else

**Review by p0mvn — APPROVED** [review]

**Comment by p0mvn** [comment]
I marked this PR as state-breaking despite backporting it so that we revisit this for a major upgrade.

The reason is that we cannot effectively test the backport since the only applications of the new `QuoTruncateMut` function are done to the new logic (post v19)

**Comment by p0mvn** [comment]
> I marked this PR as state-breaking despite backporting it so that we revisit this for a major upgrade.
> 
> The reason is that we cannot effectively test the backport since the only applications of the new `QuoTruncateMut` function are done to the new logic (post v19)

Was able to still apply this change on the backport. See commit: https://github.com/osmosis-labs/osmosis/pull/6471/commits/b1fe95c65d833cd5774cbe2534f383ca29e2bf08

Therefore, test coverage should be sufficient

**Pull request merged**